src/node_messaging.cc

```diff
@@ -544,36 +544,87 @@ std::unique_ptr<MessagePortData> MessagePort::Detach() {
 }
 
 
-void MessagePort::Send(Message&& message) {
-  Mutex::ScopedLock lock(*data_->sibling_mutex_);
-  if (data_->sibling_ == nullptr)
-    return;
-  data_->sibling_->AddToIncomingQueue(std::move(message));
-}
+Maybe<bool> MessagePort::PostMessage(Environment* env,
+                                     Local<Value> message_v,
+                                     Local<Value> transfer_v) {
+  auto isolate = env->isolate();
+  auto obj = object(isolate);
+  auto context = obj->CreationContext();
+
+  std::vector<MessagePort*> transferred_ports;
+
+  // Check if the MessagePort itself is being transferred.
+  if (transfer_v->IsArray()) {
+    Local<Array> transfer_list = transfer_v.As<Array>();
+    uint32_t length = transfer_list->Length();
+    for (uint32_t i = 0; i < length; ++i) {
+      Local<Value> entry;
+      if (!transfer_list->Get(context, i).ToLocal(&entry))
+        return Nothing<bool>();
+
+      if (env->message_port_constructor_template()->HasInstance(entry)) {
+        if (entry == obj) {
+          Local<Value> argv[] = {
+            FIXED_ONE_BYTE_STRING(isolate,
+                                  "Transfer list contains source port"),
+            FIXED_ONE_BYTE_STRING(isolate, "DataCloneError")
+          };
+          env->isolate()->ThrowException(
+              env->domexception_function()
+                  ->NewInstance(context, arraysize(argv), argv)
+                  .ToLocalChecked());
+          return Nothing<bool>();
+        }
+        MessagePort* port = Unwrap<MessagePort>(entry.As<Object>());
+        CHECK_NE(port, nullptr);
+        transferred_ports.push_back(port);
+      }
+    }
+  }
 
-void MessagePort::Send(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  Local<Context> context = object(env->isolate())->CreationContext();
   Message msg;
-  if (msg.Serialize(env, context, args[0], args[1])
-          .IsNothing()) {
-    return;
+
+  // Per spec, we need to serialize the input message even if the MessagePort
+  // is closed or detached.
+  if (data_ == nullptr) {
+    USE(msg.Serialize(env, context, message_v, transfer_v));
+    return Just(true);
```

Comment:
If we want to fulfill the `Maybe<T>` contract for this function, we should return an empty Maybe if `msg.Serialize()` returns an empty Maybe, right?